### DISCUSSION TITLE
How to find the root reason of Tcp Frame diconnect by peer (os error 104)

### DISCUSSION THREAD

**Discussion by fy2462** [tokio-rs/tokio#4339] [Q&A]
Hi guys

The problem worried me for several weeks.

I have a video frame data to send to a relay server (which follow the example/proxy.rs code), 
I create the TCP frame connection with the default `Framed<TcpStream, BytesCodec>`

1. On client A, I create a video frame message and encrypt it with the symmetric key, and then send the data with tokio TCP  frame to the relay server.
2. On the relay server-side, it only transfers the data to another client with TcpStream.
3. On client B, received the data from relay and decrypt.

But I found client A always be disconnect after sending the first 3-5 frames of video data. and output exception: `Reset by the peer`
I can confirm it is happening on network transform latency > 20ms. but I don't know why.

Could you enable any tokio log or debug info, and find out the detailed reason, for example, buffer overflow, producer or consumer speed does't match, etc.
how to check and do that. anybody help?

Thanks to all of you.

**Comment by Darksonn** [comment]
My guess is that you are framing the stream incorrectly. Are you aware that the `BytesCodec` does not have a concept of a "frame boundary", so the other end could easily receive two frames concatenated, or a single frame split into multiple smaller pieces.

**Reply by fy2462** [reply]
Thanks, @Darksonn, I am using the Frame<TcpStream, BytesCodec> on both sides(client and relay).
1. Do you have any example code to handle the situation as you know. thanks again.
2. Do you have any way to confirm with the problem you said, for example, debug log or exception output in tokio, and how to enable them.

**Reply by Darksonn** [reply]
If this is the issue, then you would need a different codec to handle it. E.g. the `LengtDelimiteredCodec`. To check it, I would probably print the lengths of your frames and check that they are as you expect.

**Reply by fy2462** [reply]
Ok, let me try. 
BTW, there is no problem when sending the video frame data directly without encryption and decryption between client and relay server, so I am very confused about the issue. anyway, I will test the solution as you said. 
Thanks.
